### PR TITLE
Add --coverage-file option

### PR DESCRIPTION
**Pull request by daejunpark**
Jenkins: test

### CODE REVIEW

**Comment by daejunpark** [comment]
Jenkins retest

**Comment by daejunpark** [comment]
Added an experimental option, `--coverage-file <file>`, which records a trace of locations of all rules and terms applied. Please review: @andreistefanescu @yilongli @dwightguth

File: java-backend/src/main/java/org/kframework/backend/java/util/Coverage.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import static org.apache.commons.io.FileUtils.writeStringToFile;
+
+/**
+ * For measuring semantic coverage
+ *
+ * @author daejunpark
+ */
+public class Coverage {
+
+    public static void print(File file, ConstrainedTerm subject) {
```
Comment (yilongli) [inline java-backend/src/main/java/org/kframework/backend/java/util/Coverage.java]: Rename `subject` to `term` or something else. Subject is only meaningful in the context of rewriting.

File: java-backend/src/main/java/org/kframework/backend/java/util/Coverage.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        if (file != null) {
+            print(file, getSourceLocation(subject));
+        }
+    }
+
+    public static void print(File file, Term subject) {
```
Comment (yilongli) [inline java-backend/src/main/java/org/kframework/backend/java/util/Coverage.java]: same as above; also add some javadoc

File: java-backend/src/main/java/org/kframework/backend/java/util/Coverage.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    private static String getSourceLocation(Rule rule) {
+        String s = null;
+        Source source = rule.getSource();
+        if (source instanceof FileSource) {
+            s = stringOf((FileSource) source) + ":" + stringOf(rule.getLocation());
+        }
+        return s;
+    }
+
+    private static String stringOf(FileSource source) {
```
Comment (yilongli) [inline java-backend/src/main/java/org/kframework/backend/java/util/Coverage.java]: rename `stringOf` to something more specific

**Comment by yilongli** [comment]
looks fine to me except for the comments above

**Comment by andreistefanescu** [comment]
lgtm

**Comment by dwightguth** [comment]
You should move the --profile flag out of MaudeKRunOptions and back into KRunOptions.Experimental, and then put this data under that flag. Like trace, the data should be output on stdout.

**Comment by dwightguth** [comment]
Can you please paste a sample of what this option outputs into this pull request so I can take a look at it? Pick something that is relatively small but showcases the features of the coverage that we want to be able to use.

**Comment by daejunpark** [comment]
For example,

$ cat a.k

```
  1 //require "builtins/int.k"
  2 
  3 module A
  4 
  5 //imports INT-HOOKS
  6 
  7 configuration <T>
  8   <k> $PGM:Cmds </k>
  9 </T>
 10 
 11 syntax Cmds ::= Cmd Cmds
 12               | Cmd
 13 syntax Cmd ::= "run1"
 14              | "run2"
 15 
 16 syntax KItem ::= "x" "(" ")"
 17                | "y" "(" ")"
 18 
 19 rule run1 => x()
 20 
 21 rule run2 => x()
 22 
 23 rule C:Cmd Cs:Cmds => C ~> Cs
 24 
 25 rule x() => .
 26 
 27 endmodule
```

$ cat 1.a

```
run1
run2
```

$ krun --coverage-file cov 1.a && cat cov

```
/Users/daejunpark/tmp2/./1.a:1:1:2:6
/Users/daejunpark/tmp2/././a.k:23:1:23:29
/Users/daejunpark/tmp2/./1.a:1:1:1:5
/Users/daejunpark/tmp2/././a.k:19:1:19:16
/Users/daejunpark/tmp2/././a.k:19:14:19:17
/Users/daejunpark/tmp2/././a.k:25:1:25:13
/Users/daejunpark/tmp2/./1.a:2:1:2:5
/Users/daejunpark/tmp2/././a.k:21:1:21:16
/Users/daejunpark/tmp2/././a.k:21:14:21:17
/Users/daejunpark/tmp2/././a.k:25:1:25:13
```

It is an alternating sequence of locations of terms and rules.

I think it needs to be stored in a separate file, because otherwise it may be interleaved with other stdout messages and make it difficult to extract the trace only.

**Comment by daejunpark** [comment]
@yilongli review again, please.

**Comment by yilongli** [comment]
lgtm

**Comment by dwightguth** [comment]
Alright, I am fine with this diff.

**Comment by dwightguth** [comment]
Jenkins: test this please

**Pull request merged**